File: sdks/js/langchain/cases/10-binary-content-redaction.js
/**
 * 10-binary-content-redaction: Binary Content Redaction
 *
 * Tests that binary image data is properly redacted in Sentry spans.
 * Sends a message with a base64-encoded image and verifies that Sentry:
 * - Captures the message structure
 * - Redacts the binary content with "[Blob substitute]" marker
 * - Does not send raw binary data to Sentry
 */

const { Sentry } = require("../setup");
const { ChatOpenAI } = require("@langchain/openai");
const { HumanMessage } = require("@langchain/core/messages");
const { runTestCase } = require("../../_test-utils/test-runner.cjs");

async function testLogic(inputs) {
  const { model, image_type } = inputs;

  const chatModel = new ChatOpenAI({
    modelName: model,
    apiKey: process.env.OPENAI_API_KEY,
  });

  // Create a small base64-encoded image (1x1 pixel PNG)
  // This is a valid 1x1 transparent PNG image
  const base64Image =
    "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==";

  // LangChain uses a specific format for multimodal messages with images
  // The content is an array with text and image_url objects
  const message = new HumanMessage({
    content: [
      {
        type: "text",
        text: "What color is this image? Answer in one word.",
      },
      {
        type: "image_url",
        image_url: {
          url: `data:image/${image_type};base64,${base64Image}`,
        },
      },
    ],
  });

  const response = await chatModel.invoke([message]);

  const text = response.content;

  if (!text) {
    throw new Error("No completion returned from LangChain");
  }

  if (process.env.SENTRY_AI_TEST_VERBOSE === "true") {
    console.log(`    Response: ${text}`);
    console.log(`    Sent message with base64-encoded ${image_type} image`);
  }
}

module.exports = runTestCase("10-binary-content-redaction", testLogic, Sentry);

File: sdks/py/langchain/cases/10-binary-content-redaction.py
"""
10-binary-content-redaction: Binary Content Redaction Test

Tests that when binary data (such as images) is sent to an LLM, Sentry
correctly redacts the binary content in the captured span data and
replaces it with a substitute marker.
"""

import os
import base64
from pathlib import Path
from langchain_openai import ChatOpenAI
from langchain_core.messages import HumanMessage
from test_runner import run_test_case


async def test_logic(inputs):
    """The actual test logic"""
    model = inputs["model"]

    chat = ChatOpenAI(model=model, api_key=os.getenv("OPENAI_API_KEY"))

    # Load static test image
    # Path: cases -> langchain -> py -> sdks -> repo_root
    repo_root = Path(__file__).parent.parent.parent.parent.parent
    image_path = repo_root / "shared" / "test-assets" / "test-image-10x10-red.png"

    with open(image_path, "rb") as f:
        image_data = f.read()
    base64_image = base64.standard_b64encode(image_data).decode("utf-8")

    # LangChain uses a specific format for image inputs
    message = HumanMessage(
        content=[
            {
                "type": "image_url",
                "image_url": {"url": f"data:image/png;base64,{base64_image}"},
            },
            {
                "type": "text",
                "text": "What color is this image? Answer in one word.",
            },
        ]
    )

    response = chat.invoke([message])

    text = response.content

    if not text:
        raise Exception("No output returned from LangChain")

    if os.getenv("SENTRY_AI_TEST_VERBOSE") == "true":
        print(f"    Response: {text}")
        print(f"    Sent image with {len(image_data)} bytes of binary data")

    return text


# Export test case functions
test_case = run_test_case("10-binary-content-redaction", test_logic)
main = test_case["main"]
assert_sentry = test_case["assert_sentry"]
